### PR TITLE
Improve write path perf

### PR DESCRIPTION
**Pull request by benraskin92**

### CODE REVIEW

File: errors/execution.go [hunk 1]
```diff
@@ -9,6 +9,9 @@ var (
 	// ErrInvalidQuery is returned when executing an unknown query type.
 	ErrInvalidQuery = errors.New("invalid query")
 
+	// ErrNilQuery is returned when trying to write a nil query
+	ErrNilQuery = errors.New("write query must not be nil")
```
Comment (nikunjgit) [inline errors/execution.go]: ErrNilWriteQuery?

File: storage/local/storage.go [hunk 2]
```diff
@@ -87,16 +89,52 @@ func (s *localStorage) Write(ctx context.Context, query *storage.WriteQuery) err
 	default:
 	}
 
+	if query == nil {
```
Comment (nikunjgit) [inline storage/local/storage.go]: do we really need this ? I think upstream should ensure this function cant get nil

File: storage/local/storage.go [hunk 3]
```diff
@@ -87,16 +89,52 @@ func (s *localStorage) Write(ctx context.Context, query *storage.WriteQuery) err
...
+	writeRequestCommon := writeRequestCommon{
+		store:      s,
+		annotation: query.Annotation,
+		unit:       query.Unit,
+		id:         id,
+	}
+
+	requests := make([]execution.Request, len(query.Datapoints))
+	for idx, datapoint := range query.Datapoints {
+		requests[idx] = newWriteRequest(&writeRequestCommon, datapoint.Timestamp, datapoint.Value)
```
Comment (nikunjgit) [inline storage/local/storage.go]: isn't this a bug if all of them share writeRequestCommon and update it in parallel ?

File: storage/local/storage.go [hunk 4]
```diff
@@ -87,16 +89,52 @@ func (s *localStorage) Write(ctx context.Context, query *storage.WriteQuery) err
...
+	return execution.ExecuteParallel(ctx, requests)
 }
 
 func (s *localStorage) Type() storage.Type {
 	return storage.TypeLocalDC
 }
+
+func (w *writeRequest) Process(ctx context.Context) error {
+	common := w.writeRequestCommon
+	store := common.store
```
Comment (nikunjgit) [inline storage/local/storage.go]: why separate only store out ?

File: services/m3coordinator/handler/write.go [hunk 5]
```diff
@@ -58,11 +59,25 @@ func (h *PromWriteHandler) parseRequest(r *http.Request) (*prompb.WriteRequest,
...
+func (w *localWriteRequest) Process(ctx context.Context) error {
+	return w.store.Write(ctx, w.writeQuery)
+}
+
+type localWriteRequest struct {
+	store      storage.Storage
+	writeQuery *storage.WriteQuery
+}
+
+func newLocalWriteRequest(writeQuery *storage.WriteQuery, store storage.Storage) execution.Request {
```
[the remark this answers is not in the document]
Comment (benraskin92) [reply services/m3coordinator/handler/write.go]: I think it reads a little better

File: storage/local/storage.go (hunk 2 above)
Comment (benraskin92) [reply storage/local/storage.go]: Yeah I wanted this in since one of our test checks for nil metrics

File: storage/local/storage.go (hunk 4 above)
Comment (benraskin92) [reply storage/local/storage.go]: Yeah- store is used for the `session` and the `namespace`, while common is used for `id`, `unit`, and `annotation`.

**Pull request merged**